viewer/packages/tools/src/SmartOverlay/SmartOverlayTool.ts

```diff
@@ -0,0 +1,400 @@
+  id: OverlayId;
+  position: THREE.Vector3;
+  infoOverlay: HTMLElement;
+};
+
+export type OverlayInfo = {
+  text: string;
+  id: OverlayId;
+  position: THREE.Vector3;
+  color?: THREE.Color;
+};
+
+export type OverlayGroupId = string;
+export type OverlayId = number;
+
+export type OverlayGroup = {
+  id: OverlayGroupId;
+};
+
+export type OverlayToolEvent = 'hover' | 'click';
+
+export type OverlayEventHandler = (event: {
+  targetOverlay: SmartOverlay;
+  mousePosition: { clientX: number; clientY: number };
+}) => void;
+
+export type SmartOverlayToolParameters = {
+  /**
+   * Max circle size in pixels. Default is 64.
+   */
+  maxCircleSize?: number;
+  /**
+   * Sets default overlay color for newly added labels.
+   */
+  defaultOverlayColor: THREE.Color;
+};
+
+export class SmartOverlayTool {
+  private readonly _viewer: Cognite3DViewer;
+  private readonly _labelIndexToLabelInfoMap = new Map<number, OverlayInfo>();
+  private readonly _overlayPoints: Overlay3DCollection[] = [];
+  private readonly _singeOverlayCollection: Overlay3DCollection;
+  private readonly _textOverlay: HTMLElement;
+
+  private readonly _defaultOverlayColor: THREE.Color = new THREE.Color('#fbe50b');
+  private readonly _temporaryVec = new THREE.Vector2();
+  private readonly _raycaster = new THREE.Raycaster();
+
+  private _isEnabled = true;
+  private _isVisible = true;
+  private _textOverlayVisible: boolean = true;
+  private _latestAddedPointIndex = 0;
+
+  private readonly _events = {
+    hover: new EventTrigger<OverlayEventHandler>(),
+    click: new EventTrigger<OverlayEventHandler>()
+  };
+
+  constructor(viewer: Cognite3DViewer, toolParameters?: SmartOverlayToolParameters) {
+    this._viewer = viewer;
+    this._defaultOverlayColor = toolParameters?.defaultOverlayColor ?? this._defaultOverlayColor;
+
+    this._textOverlay = this.createTextOverlay('', toolParameters?.maxCircleSize ?? 20);
+    viewer.domElement.appendChild(this._textOverlay);
+
+    this._singeOverlayCollection = new Overlay3DCollection();
+    this._overlayPoints.push(this._singeOverlayCollection);
+    viewer.addObject3D(this._singeOverlayCollection);
+
+    viewer.canvas.addEventListener(
+      'mousemove',
+      debounce((event: MouseEvent) => {
+        const { _textOverlay: textOverlay } = this;
+
+        const intersection = this.intersectPointsMarkers(event);
+
+        if (intersection) {
+          const { intersectedOverlay } = intersection;
+
+          const targetOverlay: SmartOverlay = {
+            text: intersectedOverlay.text,
+            id: intersectedOverlay.id,
+            position: intersectedOverlay.position,
+            infoOverlay: textOverlay
+          };
+
+          this.positionTextOverlay(event);
+          textOverlay.innerText = targetOverlay.text;
+
+          this._events.hover.fire({ targetOverlay: targetOverlay, mousePosition: event });
+        } else {
+          textOverlay.style.opacity = '0';
+        }
+      }, 10)
```

Comment:
does debounce 10 ms really have any effect?

Reply:
True, seems to be working really good without debounce, so will just remove it.